core/Tracker/Visit/VisitProperties.php

```diff
@@ -0,0 +1,58 @@
+<?php
+/**
+ * Piwik - free/libre analytics platform
+ *
+ * @link http://piwik.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ */
+
+namespace Piwik\Tracker\Visit;
+
+/**
+ * Holds temporary data for tracking requests.
+ *
+ * RequestProcessors
+ */
+class VisitProperties
+{
+    /**
+     * Information about the current visit. This array holds the column values that will be inserted or updated
+     * in the `log_visit` table, or the values for the last known visit of the current visitor.
+     *
+     * @var array
+     */
+    public $visitorInfo = array();
+
+    /**
+     * Stores plugin specific tracking request metadata. RequestProcessors can store
+     * whatever they want in this array, and other RequestProcessors can modify these
+     * values to change tracker behavior.
+     *
+     * @var string[][]
+     */
+    private $requestMetadata = array();
+
+    /**
+     * Set a request metadata value.
+     *
+     * @param string $pluginName eg, `'Actions'`, `'Goals'`, `'YourPlugin'`
+     * @param string $key
+     * @param mixed $value
+     */
+    public function setRequestMetadata($pluginName, $key, $value)
+    {
+        $this->requestMetadata[$pluginName][$key] = $value;
```

Comment:
If the metadata is stored in a core object, plugins cannot define their own metadata that can be changed by other plugins. Also the core object would have to have 'setters', since the logic to set them is accomplished through RequestProcessors, not externall.

Here is an alternative approach:

```
// in a RequestProcessor ... 
/** @var ActionsRequestMetadata $data */
$data = $request->getMetadata('Goals');
$data->goalsConverted[] = array();
```

or maybe:

```
$data = $request->getMetadata('Piwik\Goals\Tracker\GoalsRequestMetadata');

// PHPStorm will be configured to set the type of $data to the string argument
```

To avoid naming collisions and strange errors, the plugin name should be used. Setting it automatically is possible, the plugin name can be detected from the namespace.

> Eg within core (/core) we do getRequestMetadata('CoreHome', 'isVisitorKnown'), but core should not know anything about the plugins. 

This is because I only moved what was necessary for the new plugin. There is a Future TODO item about moving the rest of the logic.